addons/udes_stock/models/stock_move_line.py

```diff
@@ -366,10 +360,152 @@ def prepare(
+                    quantity_swapping, product, None, None, location
+                )
+
+                # Workout the lot_ids that were swapped in and store the ones that were swapped out
+                lot_ids_swapped_in = StockProductionLot.search(
+                    [("name", "in", lot_names_swapped_in)]
+                )
+                lot_ids_swapped_out = picking_mls.lot_id
+
+                # If any of the lot_ids_swapped_in belong to another move_line
+                swapped_lot_on_assigned_mls = StockMoveLine.search(
+                    [
+                        ("lot_id", "in", lot_ids_swapped_in.mapped("id")),
+                        ("state", "=", "assigned"),
+                        ("u_picking_type_id", "=", mls.u_picking_type_id.id),
+                        ("qty_done", "=", 0.0),
+                        ("picking_id", "!=", mls.picking_id.id),
+                    ]
+                )
+
+                # Check there is available quantity on the system
+                domain = [
+                    ("product_id", "=", product.id),
+                    ("location_id", "child_of", location.ids),
+                    ("lot_id", "in", [lot_id.id for lot_id in lot_ids_swapped_in]),
+                ]
+                quants = Quant.search(domain).with_context(prefetch_fields=False)
+                quants.read(["quantity", "reserved_quantity"], load="_classic_write")
+                if quantity_swapping > sum(quants.mapped("quantity")):
+                    raise UserError(
+                        "You have picked more stock than is available on the system. Speak to your manager."
+                    )
+
+                # Generator expressions to easily iterate over
+                mls_need_changing = (ml for ml in mls_need_changing)
+                lot_ids_swapped_in = (lot for lot in lot_ids_swapped_in)
+
+                amount_to_be_swapped = quantity_swapping
+                if is_serial:
+                    # Want to update the movelines on each serial tracked product
+                    # until all the quantity scanned in is zero
+                    qty_done = 1
+                    while quantity_swapping > 0:
+                        ml, lot_id = next(mls_need_changing), next(lot_ids_swapped_in)
+                        prod_dict = {
+                            "product_uom_qty": qty_done,
+                            "qty_done": qty_done,
+                            "lot_name": lot_id.name,
+                            "lot_id": lot_id.id,
+                        }
+                        prod_dict.update(vals)
+                        res[ml] = prod_dict
+                        quantity_swapping -= qty_done
+                else:
+                    qty_done = quantity_swapping
+                    # Update the first lot with all scanned in lot information
+                    ml, lot_id = next(mls_need_changing), next(lot_ids_swapped_in)
```

Comment:
Can we use `move_lines_for_qty()` to find the move lines we need for that lot and qty?

Reply:
The lot information we have here does not necessarily have a moveline so we can't use `move_lines_for_qty()`